designs/cloudstack-metrics-bind-addr.md

```diff
@@ -0,0 +1,176 @@
+
+cloudstack.go:
+
+#### Current:
+```
+var requiredEnvs = []string{decoder.CloudStackCloudConfigB64SecretKey}
+
+...
+func (p *cloudstackProvider) EnvMap(spec *cluster.Spec) (map[string]string, error) {
+
+	var x = spec.CloudStackDatacenter.Name
+
+	envMap := make(map[string]string)
+	for _, key := range requiredEnvs {
+		if env, ok := os.LookupEnv(key); ok && len(env) > 0 {
+			envMap[key] = env
+		} else {
+			return envMap, fmt.Errorf("warning required env not set %s", key)
+		}
+	}
+	return envMap, nil
+}
+```
+
+#### Proposed:
+```
+var requiredEnvs = []string{decoder.CloudStackCloudConfigB64SecretKey}
+var optionalEnvs = []string{decoder.CloudStackMetricsBindAddr}  // METRICS_BIND_ADDR
+
+...
+func (p *cloudstackProvider) EnvMap(spec *cluster.Spec) (map[string]string, error) {
+
+	var x = spec.CloudStackDatacenter.Name
+
+	envMap := make(map[string]string)
+	for _, key := range requiredEnvs {
+		if env, ok := os.LookupEnv(key); ok && len(env) > 0 {
+			envMap[key] = env
+		} else {
+			return envMap, fmt.Errorf("warning required env not set %s", key)
+		}
+	}
+    for _, key := range optionalEnvs {
+		if env, ok := os.LookupEnv(key); ok && len(env) > 0 {
+			envMap[key] = env
+		}
+	}
+	return envMap, nil
+}
+```
+
+#### Alternative 1:
+
+As an alternative to the provisioner passing the desired value for the metrics-bind-addr
+through an environment variable, a new parameter for this could be added to the CloudStackDatacenter struct,
+as an optional field defaulting to the secure value.
+
+As shown above, the *config.spec* is available to EnvMap() as a param, so EnvMap() could obtain the value of this new parameter from it, placing it in the environment map it's creating for running clusterctl init.  The new parameter would not be used elsewhere
+during provisioning.
+
+This would persist the override in the EKS-A Cluster Config for future reference/audit.  It would modify the
+CloudStack Datacenter struct, albeit in a non-breaking way.
+
+
+### Backwards Compatibility
+
+The proposed solution is completely backward compatible, as it makes no
+non-defaulting interface changes.  When not specified EKS-A/CAPC provision
+a cluster with the same metric-bind-addr as currently used.
+
+The Alternative Solution 1 proposes a non-breaking interface change.
+
+## User Experience
+
+
+## Security
+
+The parameter defaults to the secure option, exposing the port only within the pod.
+The parameter can be overridden by the cluster provisioner, who is making numerous
+security-impacting decisions about their clusters already.
+
+## Testing
+
+The new code will be covered by unit and e2e tests, and the e2e framework will be extended to support cluster creation across multiple Cloudstack API endpoints.
+
+The following e2e test will be added:
+
+* provision a Cloudstack cluster with metrics-bind-addr overridden.  Confirm that the resulting cluster's
+CAPC manager pod is configured with this address.
```

Comment:
what does this confirmation look like? Currently we're able to use the kubectl client in eks-a to query eks-a resources. I'm not sure how much we'd like to tweak that to fetch and assert on provider-specific resources, like the capc manager pod. Shouldn't a CAPC e2e test, combined with EKS-A unit tests be sufficient? Otherwise it feels like we're testing a little too deeply from eks-a's perspective.